src/pip/_internal/operations/prepare.py

```diff
@@ -464,28 +447,68 @@ def _http_get_download(session, link):
     return resp
 
 
+class Download(object):
+    def __init__(
+        self,
+        response,  # type: Response
+        link,  # type: Link
+        filename,  # type: str
+        chunks,  # type: Iterable[bytes]
+    ):
+        # type: (...) -> None
+        self.response = response
+        self.link = link
+        self.filename = filename
+        self.chunks = chunks
+
+
+class Downloader(object):
+    def __init__(
+        self,
+        session,  # type: PipSession
+        progress_bar,  # type: str
+    ):
+        # type: (...) -> None
+        self._session = session
+        self._progress_bar = progress_bar
+
+    def __call__(self, link):
+        # type: (Link) -> Download
+        try:
```

Comment:
The `redacted_url` (currently computed in `_prepare_download`) could be computed here and passed to `_prepare_download` instead of the `link`: it could then also be used in this `logger.critical`.

Reply:
`Link` does `redact_auth_from_url` in its `__str__` [here](https://github.com/pypa/pip/blob/master/src/pip/_internal/models/link.py#L69-L78), and IMO the simplification of the URL in `_prepare_download` makes debugging harder (just one more detail to remember about our log output).

Reply:
Created #7390 to discuss it.